panic: unexpected keypair dbname='XXXX' from pq

Can't start freshly installed grafana agent. Got this error while starting the service:

```
panic: unexpected keypair dbname='xxxx' from pq
Nov 18 22:11:48 ip-10-0-x-xx grafana-agent[13306]: goroutine 1 [running]:
Nov 18 22:11:48 ip-10-0-x-xxx grafana-agent[13306]: github.com/grafana/agent/pkg/integrations/postgres_exporter.parsePostgresURL(0xc00070c140, 0x4d, 0x1, 0x1, 0x0)
Nov 18 22:11:48 ip-10-0-x-xxx grafana-agent[13306]:         /drone/src/pkg/integrations/postgres_exporter/postgres_exporter.go:90 +0x2eb
Nov 18 22:11:48 ip-10-0-x-xxx grafana-agent[13306]: github.com/grafana/agent/pkg/integrations/postgres_exporter.(*Config).InstanceKey(0xc000125520, 0xc000945d88, 0x13, 0x2, 0x2, 0xc000945d88, 0x13)
Nov 18 22:11:48 ip-10-0-x-xxx grafana-agent[13306]:         /drone/src/pkg/integrations/postgres_exporter/postgres_exporter.go:58 +0x85
Nov 18 22:11:48 ip-10-0-x-xxx grafana-agent[13306]: github.com/grafana/agent/pkg/integrations.(*Manager).ApplyConfig(0xc000950ea0, 0x1, 0xc000dd2fb0, 0x1, 0x1, 0x0, 0xc0005a00a8, 0x1, 0x1, 0x12a05f200, ...)
Nov 18 22:11:48 ip-10-0-x-xxx grafana-agent[13306]:         /drone/src/pkg/integrations/manager.go:252 +0x1939
Nov 18 22:11:48 ip-10-0-x-xxx grafana-agent[13306]: github.com/grafana/agent/pkg/integrations.NewManager(0x1, 0xc000dd2fb0, 0x1, 0x1, 0x0, 0xc0005a00a8, 0x1, 0x1, 0x12a05f200, 0x3039, ...)
Nov 18 22:11:48 ip-10-0-x-xxx grafana-agent[13306]:         /drone/src/pkg/integrations/manager.go:182 +0x1f8
Nov 18 22:11:48 ip-10-0-x-xxx grafana-agent[13306]: main.NewEntrypoint(0xc000d73890, 0xc00096a000, 0xc000dd2b10, 0xc000dd3cf0, 0x4419e5, 0xc000076710)
Nov 18 22:11:48 ip-10-0-x-xxx grafana-agent[13306]:         /drone/src/cmd/agent/entrypoint.go:93 +0x54a
Nov 18 22:11:48 ip-10-0-x-xxx grafana-agent[13306]: main.main()
Nov 18 22:11:48 ip-10-0-x-xxx grafana-agent[13306]:         /drone/src/cmd/agent/main.go:66 +0x1d
```

Sample configuration `/etc/grafana-agent.yml`:

```
integrations:
  postgres_exporter:
    data_source_names:
    - "postgresql://linus:42secret@localhost:5432/postgres?sslmode=disable"
    enabled: true
    
    ......
```

Seems there are bugs with the `postgres_exporter.parsePostgresURL()`

tried to reproduce in this golang playground got the same result: https://play.golang.org/p/oEukXPOHXvc

--- a/CHANGELOG.md
+++ b/CHANGELOG.md
@@ -1,5 +1,7 @@
 # Main (unreleased)
 
+- [BUGFIX] Fix panic when using postgres_exporter (@saputradharma)
+
 # v0.21.0 (2021-11-17)
 
 - [ENHANCEMENT] Update Cortex dependency to v1.10.0-92-g85c378182. (@rlankfo)
@@ -23,8 +25,6 @@
 - [BUGFIX] Operator: Fix relabel_config directive for PodLogs resource (@hjet)
 
 - [BUGFIX] Traces: Fix `success_logic` code in service graphs processor (@mapno)
-
-- [BUGFIX] Fix panic when using postgres_exporter (@saputradharma)
 
 - [CHANGE] Self-scraped integrations will now use an SUO-specific value for the `instance` label. (@rfratto)
 

--- a/CHANGELOG.md
+++ b/CHANGELOG.md
@@ -24,6 +24,8 @@
 
 - [BUGFIX] Traces: Fix `success_logic` code in service graphs processor (@mapno)
 
+- [BUGFIX] Fix panic when using postgres_exporter (@saputradharma)
+
 - [CHANGE] Self-scraped integrations will now use an SUO-specific value for the `instance` label. (@rfratto)
 
 - [CHANGE] Traces: Changed service graphs store implementation to improve CPU performance (@mapno)

--- a/pkg/integrations/postgres_exporter/postgres_exporter.go
+++ b/pkg/integrations/postgres_exporter/postgres_exporter.go
@@ -85,7 +85,7 @@ func parsePostgresURL(url string) (map[string]string, error) {
 	res := map[string]string{}
 
 	for _, keypair := range strings.Split(raw, " ") {
-		parts := strings.SplitN(keypair, "=", 1)
+		parts := strings.Split(keypair, "=")
 		if len(parts) != 2 {
 			panic(fmt.Sprintf("unexpected keypair %s from pq", keypair))
 		}

--- a/pkg/integrations/postgres_exporter/postgres_exporter.go
+++ b/pkg/integrations/postgres_exporter/postgres_exporter.go
@@ -84,13 +84,24 @@ func parsePostgresURL(url string) (map[string]string, error) {
 
 	res := map[string]string{}
 
+	unescaper := strings.NewReplacer(`\'`, `'`, `\\`, `\`)
+
 	for _, keypair := range strings.Split(raw, " ") {
-		parts := strings.Split(keypair, "=")
+		parts := strings.SplitN(keypair, "=", 2)
 		if len(parts) != 2 {
 			panic(fmt.Sprintf("unexpected keypair %s from pq", keypair))
 		}
 
-		res[parts[0]] = parts[1]
+		key := parts[0]
+		value := parts[1]
+
+		// Undo all the transformations ParseURL did: remove wrapping
+		// quotes and then unescape the escaped characters.
+		value = strings.TrimPrefix(value, "'")
+		value = strings.TrimSuffix(value, "'")
+		value = unescaper.Replace(value)
+
+		res[key] = value
 	}
 
 	return res, nil

--- a/pkg/integrations/postgres_exporter/postgres_exporter_test.go
+++ b/pkg/integrations/postgres_exporter/postgres_exporter_test.go
@@ -1,0 +1,24 @@
+package postgres_exporter
+
+import (
+	"reflect"
+	"testing"
+)
+
+func Test_ParsePostgresURL(t *testing.T) {
+	dsn := "postgresql://linus:42secret@localhost:5432/postgres?sslmode=disable"
+	expected := map[string]string{
+		"dbname":   "'postgres'",
+		"host":     "'localhost'",
+		"password": "'42secret'",
+		"port":     "'5432'",
+		"sslmode":  "'disable'",
+		"user":     "'linus'",
+	}
+
+	actual, _ := parsePostgresURL(dsn)
+	if !reflect.DeepEqual(actual, expected) {
+		t.Errorf("parsePortgresURL failed, actual: %v, expected: %v", actual, expected)
+	}
+
+}

--- a/pkg/integrations/postgres_exporter/postgres_exporter_test.go
+++ b/pkg/integrations/postgres_exporter/postgres_exporter_test.go
@@ -1,8 +1,9 @@
-package postgres_exporter
+package postgres_exporter //nolint:golint
 
 import (
-	"reflect"
 	"testing"
+
+	"github.com/stretchr/testify/require"
 )
 
 func Test_ParsePostgresURL(t *testing.T) {
@@ -16,9 +17,8 @@ func Test_ParsePostgresURL(t *testing.T) {
 		"user":     "linus",
 	}
 
-	actual, _ := parsePostgresURL(dsn)
-	if !reflect.DeepEqual(actual, expected) {
-		t.Errorf("parsePortgresURL failed, actual: %v, expected: %v", actual, expected)
-	}
+	actual, err := parsePostgresURL(dsn)
+	require.NoError(t, err)
+	require.Equal(t, actual, expected)
 
 }

--- a/pkg/integrations/postgres_exporter/postgres_exporter_test.go
+++ b/pkg/integrations/postgres_exporter/postgres_exporter_test.go
@@ -8,12 +8,12 @@ import (
 func Test_ParsePostgresURL(t *testing.T) {
 	dsn := "postgresql://linus:42secret@localhost:5432/postgres?sslmode=disable"
 	expected := map[string]string{
-		"dbname":   "'postgres'",
-		"host":     "'localhost'",
-		"password": "'42secret'",
-		"port":     "'5432'",
-		"sslmode":  "'disable'",
-		"user":     "'linus'",
+		"dbname":   "postgres",
+		"host":     "localhost",
+		"password": "42secret",
+		"port":     "5432",
+		"sslmode":  "disable",
+		"user":     "linus",
 	}
 
 	actual, _ := parsePostgresURL(dsn)

